\documentclass{book}

\usepackage[english]{babel}
\usepackage{listings}

% for tikz images
\usepackage[x11names, rgb]{xcolor}
\usepackage[utf8]{inputenc}
\usepackage{tikz}
\usetikzlibrary{snakes,arrows,shapes}
\usepackage{amsmath}


\newcommand\ensembl{EnsEMBL }
\newcommand\CLCB{CLCB }
\newcommand\lisp{Lisp }
\newcommand\eg{e.\,g.}


\title{Common Lisp Computational Biologiy - Handbook}

\begin{document}

\titlepage
\tableofcontents

\chapter{Introduction}


\section{Motivation for \CLCB}

Lisp is fun. And it was felt that for a particular problem at hand, 
there was little gained by using Bio* libraries for Java, Python or Perl.
Hence, the decision was made to implement in Lisp and from day one kept in
mind to offer the fun to others, too.

One can program Lisp as if it was Perl, which is referred to as
''declarative''. What Lisp once was special for is the {\em fun}ctional
programming. Everyone has experienced what functional programming is alike
who has specified the first argument to the {\em sort} routine in Perl,
which is the expression deciding if the objects \$a or \$b is larger, with
a non-trivial expression.  In Lisp,
such functions that do not need a name are called $\lambda$ expressions.
One takes such, optionally gives them a name to make them real functions,
and applies them to lists (or another structure) of objects. To the
biologically minded in us, such functions are much like an enzyme
that is acting on some sequence of nucleotides. In other languages
one only ''calls'' functions by name. In Lisp, functions are something
tangible. Admittedly, these other languages do offer constructs that
allow for a functional programming to some degree. But with Lisp it is
the other way around.

But exactly {\em why} is it so much fun? There are probably many answers
to this questions.  For one, the language is very clean, even for concepts
that came into fashion twenty years after Lisp's conception. Take
for instance the overloading of methods when introducing a subclass.
When interating over a set of zoo animals, there is no need for the
typical switch statement to learn about the object's real type and then a
cast to call the real method. If in C or Java you would do something alike

\lstset{language=C}
\begin{lstlisting}
for(animal *a= alist->first(); a->next();) {
	if ("dog"==typeof(a)) {
		dog *d = (dog *)a;
		d->speak();
	}
	else if ("cat"==typeof(a)) {
	...
	}
}
\end{lstlisting}

to hear the miows and barks.  In Lisp, you'd get the sound directly for
any animal, also for the ones not originally anticipated in that zoo loop.
For bioinformatics, with its zoo of sequence types and their subtypes,
the genes and what was formerly known as "junk", such cleanliness is essential.

There are other reasons $\ldots$

\section{How to learn Lisp}

The Net is full of tutorials that introduce to the language
Lisp. We found the instructions at the GMU very good to follow
(http://www.cs.gmu.edu/~sean/lisp/LispTutorial.html). A much recommended
text book is

	Peter Seibel\\
	Practical Common Lisp\\
	ISBN 1-59059-239-5\\
	Apress

A more biologically inclined reader of this document may decide
to leave the syntax of the language aside while first passing through
this document and instead concentrate more on the understanding
of the classes that represent biological entities in \CLCB.

\section{Instant access to \CLCB without reading this documentation}

\CLCB is developed with SBCL but should be compatible with other
implementations with Lisp. This section gives you a head start into the
code. Frankly, the real documentation is in the code, literally, this is
where '':documentation'' slots allow for the annotation of what was
programmed and we prepare to parse that information from the code into
this document. But this is only now being worked on.
Unless you have a colleague who is already proficient in \CLCB, 
there is no alternative to jumping right in.

\subsection{Setting up the interpreter}

In your UNIX shell change the directory to where the README 
of \CLCB is. The UNIX command ''pwd'' informs about the current
working directory. Then start the Lisp interpreter, possibly just
by typing ''sbcl''. The interpreted now needs to learn about 
where to find your \CLCB source code. To achieve that, in the lines below
substitute ''path/to/CLCB'' with that the command ''pwd'' just sent to
the screen and type it into the interpreter. You'd need to do this
every time you start up. Your interpreter will have files it reads at
startup and you should in the longer run move these lines there.

\lstset{language=lisp}
\begin{lstlisting}
(pushnew '<path/to/CLCB> asdf:*central-registry*)
(pushnew '<path/to/CLCB/ensembl> asdf:*central-registry*)
(asdf:oos 'asdf:load-op :clcb)
(asdf:oos 'asdf:load-op :clcb-ensembl)
(in-package :ensembl)
\end{lstlisting}

The clcb-utils library is loaded automatically since clcb-ensembl depends on it.

\subsection{First hands on experience}

When looking at the examples given below you need to keep in mind that
computer scientists try hard to distinguish between the data and the
way it is presented to the user. The reason is that the same data
is interpreted differently in different contexts and there is hence
no way to know for sure how the data shall be shown to meet the user's
expectations. The examples below give references to the data, indicating
that all data is available to the programmer. This is great to the
programmer, saving lots of work. An abstract way to present the data,
i.e, as LaTeX tables, ASCII with white space or tab delimited, is still
being designed.

The code below retrieves a gene from \ensembl \cite{BiAnCaChClCoCoCuCuCuDoDuFeFlGrHaHeHoHu06}, links to its its
transcripts and finally the first transcript's protein product. The
asterisk ($*$) is the prompt of the sbcl shell. If executed from within
the Emacs smile environment, then the prompt is the name of the current
package followed by the greater sign ($>$).  The commands at the prompt
always start with an open parathesis and end with a matching closing
one. The output appears right below the input and is sometimes not
straight forward to distiguish from the when it was not entered by
oneself.  Be aware that the translation process uses ''*'' to refer to
the result of the previous command, not as a prompt. In the following,
a commented interaction with the \lisp shell shown. It starts with the
retrieval of a gene from \ensembl by its stable gene ID.

\begin{lstlisting}
* (fetch-by-stable-id ``ENSG00000000005'')
#<GENE {C974EC1}>
\end{lstlisting}

The gene is retrieved. The object itself is not directly shown to the
screen as it contains quite a lot of internal inforamtion. Instead, by
an \lisp-intrinsic automatism, the macro print-object is called which
delivers these neat $\#<\ldots>$ construct. The contents themselves can
be passed to a global variable $*gene*$.

\begin{lstlisting}
(defparameter *gene* (fetch-by-stable-id ``ENSG00000000005''))
\end{lstlisting}

Other routines can perform transformations of this data, i.e. the retrieval of
\begin{itemize}
\item transcripts,

\begin{lstlisting}
* (transcripts *gene*)
(#<TRANSCRIPT {CAB4751}>)
\end{lstlisting}

\item translations or

\begin{lstlisting}
* (translation (car *))
#<TRANSLATION {CAF4C49}
\end{lstlisting}

\item exons. Below, only exons that contribute to the first transcript are retrieved:
\begin{lstlisting}
* (exons (first (transcripts *gene*)))
(#<EXON {CB4F9E9}> #<EXON {CB5C809}> #<EXON {CB69539}>
 #<EXON {CB76249}> #<EXON {CB831E9}> #<EXON {CB8FF09}>
 #<EXON {CB9CC09}> #<EXON {CBA98E9}> #<EXON {CBB6619}>
 #<EXON {CBC3329}> #<EXON {CBD0011}> #<EXON {CBDCD09}>
 #<EXON {CBE9A09}> #<EXON {CBF6729}> #<EXON {CC03421}>
 #<EXON {CC10149}> #<EXON {CC1CE61}> #<EXON {CC2BB61}>
 #<EXON {CC3A8A9}> #<EXON {CC47AD1}> #<EXON {CC56809}>)
\end{lstlisting}
\end{itemize}

The retrieval of features can also be applied to protein sequences. Features of
genomic sequences are transcribed units, those of transcribed units could be the
gene structure. Of proteins, it is protein domains of various sorts or
post-translational modifications.

\begin{lstlisting}
* (protein-features (translation (car (transcripts *gene*))))
(#<PROTEIN-FEATURE {CD15729}> #<PROTEIN-FEATURE {CD178C9}>
 #<PROTEIN-FEATURE {CD18591}> #<PROTEIN-FEATURE {CD19269}>
 #<PROTEIN-FEATURE {CD19F31}> #<PROTEIN-FEATURE {CD1ABF9}>
 #<PROTEIN-FEATURE {CD1B8C9}> #<PROTEIN-FEATURE {CD1C591}>)
\end{lstlisting}

These featues are of different types, which can be queried individually:

\begin{lstlisting}
* (mapcar #'protein-feature-type
          protein-features
	    (translation
	      (car (transcripts *gene*)))))
("Pfam" "Superfamily" "Superfamily" "Smart" "low_complexity"
"low_complexity" "low_complexity" "Prosite_profiles")
\end{lstlisting}

Once the protein-features are available, \CLCB offers to 
present their genomic coordinates. This way, one gets the
exact bases that sign responsible for, say, the interface
region of an receptor to a ligand.

\begin{lstlisting}
(mapcar #'dna-sequence-interval
       ((protein-features
          (translation
	     (car (transcripts *gene*)))))
(#<DNA-SEQUENCE {D1DF909}>
 #<MULTI-INTERVAL (#<DNA-SEQUENCE {D1FD389}> #<DNA-SEQUENCE {D1FE011}>)>
 #<MULTI-INTERVAL (#<DNA-SEQUENCE {D21BEF1}> #<DNA-SEQUENCE {D21CB71}>
                   #<DNA-SEQUENCE {D21D7F1}>)>
 #<DNA-SEQUENCE {D2D2A01}> #<DNA-SEQUENCE {D2EA241}> #<DNA-SEQUENCE {D3019E1}>
 #<MULTI-INTERVAL (#<DNA-SEQUENCE {D320179}> #<DNA-SEQUENCE {D320DF9}>)>
 #<DNA-SEQUENCE {D342429}>)
\end{lstlisting}

But, something appears strange, it is abstract objects that define the sequence.
The coordinates are
not directly visible. All stretches of the genome are presented
as intervals, something that BioPerl refers to as a slice. But here
in \CLCB we have all the algebraic operations defined on these intervals,
which the BioPerl slices (to our knowledge) do not know about. This way,
you can, \eg,  intersect with other intervals like exons or intergenomic
similarities.  One can also build the complements. Frankly, until the
very end of the analysis, we do not give much about knowing the exact
numbers of the location.  But then, we can have it:

\begin{lstlisting}
* (mapcar #'(lambda (x)
                    (genomic-coordinates (dna-sequence-interval x)))
            (protein-features
	       (translation (car (transcripts *gene*)))))
((204639 . 204885) (204630 . 205434) (250229 . 255355)
 (204639 . 204918) (242551 . 242602) (216822 . 216855)
 (204921 . 205473) (204639 . 204888))
\end{lstlisting}

But where is the chromosome specified and where is the organism? Well,
normally you know about these already and they are invariant whenever
one is inspecting a particular protein that is not a product of a
translocation $\ldots$ hm $\ldots$. It is the gene that can
be queried for the chromosome.

\begin{lstlisting}
* (chromosome *gene*)
"18"
\end{lstlisting}

\section{Acknowledgements}

\chapter{Packages of \CLCB}

\section{molecule}

The molecule class can be understood as the root of the biological interpretation of the data retrieved, much in the same way that the interval classes are the root of the analyses.

\begin{figure}
\begin{centering}
\input{moleculeClassDiagram}
\end{centering}
\end{figure}

\section{bio-sequence}

Conceptionally, the bio-sequence class is the link between the abstract mathematical notion of an interval and the very hands-on data that one retrieves from Ensembl or possibly elsewhere.

\begin{figure}
\begin{centering}
\input{biosequenceClassDiagram}
\end{centering}
\end{figure}

The package describes several routines for the input and output of sequences to streams.

\section{ensembl}

The \ensembl database provides insights into 
\begin{itemize}
\item genomic sequences,
\item genes,
\item transcripts,
\item proteins,
\item sequence variation and
\item intergenomic comparisons.
\end{itemize}
The initiative comes with a Perl-generated web site, an \ensembl
Perl API and was instrumental for the development of BioPerl
\cite{bioperl:2002}. So, if they are all using Perl, why Lisp? Well,
\ensembl uses an old release of BioPerl and has experimented with a Java
interface \ensembl, which was dumped again. For an easier access to the
data on the database level, the effort BioMart was initiated.

BioMart takes the core data of \ensembl and prepares views on the data that
facilitate dramatically the preparationof everyday's analyses. However,
for anything out of the row, it is not prepared. The \ensembl classes
presented here take a different approach. They follow the same logic of
BioMart and some of the \ensembl API classes to retrieve data from the core
tables. Other than BioMart, all interactions are dynamically determined
and new interactions are evolving just as they are needed. Other than
BioPerl, only minimal code is needed for that effort and the results
are immediately accessible to further analyses. Add to that all the
differences that Lisp makes by itself.

\subsection{Preparation of \ensembl classes}

The classes in the clcb-ensembl package basically represent the tables of
the \ensembl core MySQL tables. An instance of that class shall represent 
a single entry in the table of the same name. With possibly a few transient
exceptions, the unique identifiers of a row have a single attribute with the 
name which is a concatenation of the table name and the suffix 'id'. Other
tables refer to entries of the prior table with an attribute of the same name.
Consequently, the formulation of queries in Ensembl databases seems straight
forward.

Lisp has the flexibility to utilise that straight-forwardness with
macros. One passes them a table name, the attributes, and the Lisp macros
will figure out themselves what the class definition should look alike.
In a way, at least for the standard attributes, it would not have benn required to
specify the attributes manually since these could be read from the database dynamically.
But it is fine as it is already.

\subsubsection{Preparation of views on Ensembl}

The first macro automates the specification views on the Ensembl data. These
can be understood as resulting from an arbitrary SQL query and much like in 
SQL, the results get the character of a class themselves.

\begin{lstlisting}
(defmacro def-ensembl-view (name superclasses slots &rest options)
  "Define an EnsEMBL class."
  (let ((supers (remove-duplicates (append superclasses '(ensembl-object))))
        (cl-options
         (nconc (list :base-table (substitute #\_ #\- (string-downcase name)))
                options)))
    `(def-view-class ,name ,supers ,slots ,cl-options)))
\end{lstlisting}

Below you see the macro applied to specify an abstract view on DNA sequence data:
\begin{lstlisting}
(def-ensembl-view dna-sequence ()
  ((seq-region-id :type integer
                  :db-constraints :not-null
                  :initarg :seq-region-id)
   (seq-region-start :type integer
                     :db-constraints :not-null
                     :initarg :seq-region-start)
   (seq-region-end :type integer
                   :db-constraints :not-null
                   :initarg :seq-region-end)
   (seq-region-strand :type integer
                      :db-constraints :not-null
                      :initarg :seq-region-strand)
   (seq-region :accessor seq-region
               :db-kind :join
               :db-info (:join-class seq-region
                         :home-key seq-region-id
                         :foreign-key seq-region-id
                         :set nil))
   (previous-element-fn :accessor previous-element-fn
                        :initform #'1-
                        :db-kind :virtual
                        :allocation :class)
   (next-element-fn :accessor next-element-fn
                    :initform #'1+
                    :allocation :class
                    :db-kind :virtual)
   (element-member-fn :accessor element-member-fn
                      :initform
                      #'(lambda (set x)
                          (with-slots (seq-region-start seq-region-end) set
                            (and (integerp x)
                                 (<= seq-region-start x seq-region-end))))
                      :allocation :class
                      :db-kind :virtual)))
\end{lstlisting}

And for more 1:1 links between classes and tables, the def-ensembl-class is available.

\begin{lstlisting}
(defmacro def-ensembl-class (class-name slots &key dna-sequence stable-id-char)
  (with-symbol-prefixing-function (table-concat class-name)
    `(progn
       ,(when stable-id-char
              `(progn
                (def-ensembl-stable-id-view ,class-name)
                (def-fetch-by-stable-id-method ,class-name)
                (push (cons ,stable-id-char ',class-name) *char-class-alist*)))
       (def-ensembl-view ,class-name (,@(when dna-sequence '(dna-sequence)))
          ,(append ;; If the object has a stable id, we need to link
                   ;; to the appropriate table.
                   (when stable-id-char
                     `((stable-id
                        :db-kind :join
                        :db-info (:join-class ,(table-concat "-STABLE-ID")
                                  :home-key ,(table-concat "-ID")
                                  :foreign-key ,(table-concat "-ID")
                                  :set nil))))
                   ;; The slot definitions for the table.
                   slots)))))
\end{lstlisting}

The lines below show the representation of the exon class:
\begin{lstlisting}
(def-ensembl-class exon
  ((exon-id :type integer
            :db-type :key
            :db-constraints (:primary-key :not-null))
   (phase :type integer
          :accessor exon-phase)
   (end-phase :type integer
              :accessor exon-end-phase)
   (is-current :type integer)
   (exon-transcript :db-kind :join
                    :db-info (:join-class exon-transcript
                              :home-key exon-id
                              :foreign-key exon-id
                              :set nil)))
  :stable-id-char #\e
  :dna-sequence t)
\end{lstlisting}

\section{utils}
\subsection{intervals}

A core principle of \CLCB is to retrieve data from Ensembl as easily as
possible and then continue with the analysis with minimal and transparent
access to the database. On major technical component to achieve such is
by the introduction of intervals to represent subsequences.

\begin{figure}
\begin{center}
\input{intervalClassDiagram}
\end{center}
\caption{{\bf Class diagram of intervals in \CLCB.} }
\label{fig:intervalClassDiagram}
\end{figure}

The notion of an interval is very intuitive both to users from a computational
or a biological background. Figure \ref{fig:intervalClassDiagram} displays
the relationships between these classes. As a start, we need only data intervals
on natural numbers. Later, we will also have negative numbers, i.e., to explicitly
include promotor regions. For the moment this is not implemented since in biology,
it is common to omit position 0. The interval of $[ -1, -1 ]$ then includes two numbers,
not three as the computationally primed user might expect.

To access the classes and functions defined for the intervals, one needs to attach onselves to the clcb-utils package:

\begin{lstlisting}
* (in-package :clcb-utils)

#<PACKAGE "CLCB-UTILS">
\end{lstlisting}


To generate intervals, specify the type of interval at hand and specify its borders
\begin{lstlisting}
* (defparameter *interval1*
  (make-interval 'integer-interval 0 100))

*INTERVAL1*
\end{lstlisting}

and to inspect its contents

\begin{lstlisting}
* *INTERVAL1*

#<INTEGER-INTERVAL [0,100]>
\end{lstlisting}

The function mapcar applies a function consecutively to all values of a list of attributes. 
In the case below, the mapcar expression receives two lists as arguments and the lambda
expression expects two values from the lists, like
\begin{lstlisting}
* (mapcar (lambda (x y) (list x y)) '(1 2 3) '(1 2 3))

((1 1) (2 2) (3 3))
\end{lstlisting}

Analogously, in the example below, the quoted lists define one multi-intervals $[5;13]\cup[50;60]\cup[70;90]\cup[65;70]$.

\begin{lstlisting}
* (defparameter *list-of-intervals*
               (mapcar (lambda (start end)
		               (make-interval
			          'integer-interval
			          start end))
	               '(5  50 70 65)
           	       '(13 60 90 70)))
*LIST-OF-INTERVALS*
\end{lstlisting}

The result was hidden away in a variable to simplify
the later expressions. To inspect it:

\begin{lstlisting}
* *LIST-OF-INTERVALS*

(#<INTEGER-INTERVAL [5,13]> #<INTEGER-INTERVAL [50,60]>
 #<INTEGER-INTERVAL [70,90]> #<INTEGER-INTERVAL [65,70]>)
\end{lstlisting}

To work comfortably with such sets of intervals, a separate
aggregator class was defined that takes the list as an input
for its constructor:

\begin{lstlisting}
* (defparameter *multi-interval*
     (make-instance 'multi-interval :intervals *LIST-OF-INTERVALS*))
*MULTI-INTERVAL*
* *MULTI-INTERVAL*

#<MULTI-INTERVAL (#<INTEGER-INTERVAL [5,13]> #<INTEGER-INTERVAL [50,60]> #<INTEGER-INTERVAL [65,90]>)>

\end{lstlisting}

An application of intervals is on genomic sequences. 

\begin{lstlisting}
dna-sequence-interval
\end{lstlisting}



\chapter{Use Cases}

This section introduces to a couple of szenarios in which \CLCB may
be handy. It is probably mostly evident that the functionality of CBCL
is not complete. And it will never be complete. But it will always be
nice. The challenge is hence to use it at its best and there will always
be a traitoff between staying within Lisp and coming up with a mixture
with related solutions in other languages.

\section{Browsing \ensembl core database}

In its current state, the CBCL is ahead of its time in its interaction with
the Ensembl MySQL databases - but nowhere else, really. For large analyses
it seems like a sensible idea to have a local mirror of the data since
\CLCB (still) comes with a fairly high number of queries. The connections
remain open between these queries.

\subsection{Link between genomic and protein features}

\subsection{Selection of alternatively spliced protein features}

\subsection{Tabular representation of findings}

\section{Intergenomics}

\subsection{\ensembl Mart}

{\em Implemented while writing these lines.}

\subsection{\ensembl Compara}

Not implemented yet.

\section{Communication}

The information that was gathered with \CLCB most
likely needs to be processed further. This may
be for publication or to add more biological data
that was not directly accessible from \CLCB. Also,
some external parties may have already prepared 
an analysis platform that should be sent the data to.

\subsection{Data import}

Any data can be read with Lisp. The clcb-utils:table
class is of considerable benefit to add data for analyses.
The files amino-acids.\{txt,lisp\} and species.\{txt,lisp\}
give examples for the preparation of a hash table
that finds rows in tables by some attribute of theirs.

To prepare the input, a new class that is representing the rows must
be prepared. In a initial version, it may be the names of the columns
that directly determine the names of the slots. At a later stage on may
decide that some column are not of interest or that additional slots
are required.

The genetic code representation represents a custom format.  If the data
imported is likely to be of interest for a larger audience then please
consider to submit your work to the \CLCB or make it publicly available
through other means.

\subsubsection{Output text files and tables}

The output of tables to prepare a text-based communication with another
tool can work in complete analogy to the input. One defines a table and
just by printing it to a file this will be directly readable by, say,
R, Gnumeric or some other spreadsheet.

If the output format is not tab-delimited or in any way repetitive, e.g.,
a dendrogram, then core Lisp may be the perfect handler of that format,
but not the routines that are currently provided by \CLCB.

\subsubsection{Relational databases}



\subsection{Retrieval of database cross references}

implementiert Steffen mal

\subsection{...with other Bioinformatics Suites}

At this developmental stage, \CLCB has not yet developed any particular
interface to other bioinformatics libraries or statistics suites. To
tap into all the resources made available to R (www.r-project.org) seem
a worthwhile target. To achieve such, the data needs to be exchanged
between Lisp and R as text files.

The orchestration of software programs to work towards the scientific
or very pragmatic aims is an art per se. Tools like Taverna have become
instrumental in this respect but we have not yet found the opportunity
to provide an interface to this workflow suite. \CLCB should only be
use for what it can do best, and this is the communication with Ensembl
to retrieve data. Once the data is available, one needs to decide for
oneself if the downstream analyses should become part of \CLCB or if
there is another tool that is performing that task already. If so,
then CBCL should already allow to store the data in a format that is
suitable as input for later tools since the most common file formats
are already implemented.

The tool that is most commonly used to perform the orchestration is the
program {\em make}. The data files that every step of the analysis needs
are communication and synchronisation points. Only after the process
producing one data file has been completed, the downstream analyis
programs are initiated.  Also nice: Makefiles can implement a parallel
execution of the analysis and even some queuing systems are compatible
with Makefiles to allow using a complete compute farm or a computational
grid for the analysis.

There is some ever-going work on applying agent technologies for
Bioinformatics \cite{agents:2007}. Well, today this is mostly written
in Java, but Lisp of course has a long lasting history in this field.

\bibliographystyle{plain}
\bibliography{clcb}

\end{document}
